File: java/Dynamic Programming/MaximumSubarraySumDpOptimized.java
public class MaximumSubarraySumDpOptimized {
    public int maximumSubarraySumDpOptimized(int[] nums) {
        int n = nums.length;
        if (n == 0) {
            return 0;
        }
        int currentSum = nums[0];
        int maxSum = nums[0];
        for (int i = 1; i < n; i++) {
            currentSum = Math.max(nums[i], currentSum + nums[i]);
            maxSum = Math.max(maxSum, currentSum);
        }
        return maxSum;
    }
}

File: kotlin/Dynamic Programming/MaximumSubarraySumDPOptimized.kt
fun maximumSubarraySumDPOptimized(nums: IntArray): Int {
    val n = nums.size
    if (n == 0) {
        return 0
    }
    var currentSum = nums[0]
    var maxSum = nums[0]
    for (i in 1 until n) {
        currentSum = maxOf(nums[i], currentSum + nums[i])
        maxSum = maxOf(maxSum, currentSum)
    }
    return maxSum
}
